test/test_core.py

```diff
@@ -97,8 +97,8 @@ def test_from_import_non_existing_julia_name(self):
     def test_julia_module_bang(self):
         from julia import Base
         xs = [1, 2, 3]
-        ys = Base.scale_b(xs[:], 2)
-        assert all(x * 2 == y for x, y in zip(xs, ys))
+        ys = Base.fill_b(xs[:], 10)
```

Comment:
Why not `ys = Base.fill_b(xs, 10)` ?

Reply:
I realize that, but the new test doesn't actually test much.  That's why I suggested just setting `ys = [2, 4, 6]` and keeping the test of `all` as-is.

Reply:
My intention when writing the test was just to test variable renaming code path (back then I wasn't sure if it was OK to import numpy).  But now that we have it via tox, I suggest something like:

```python
xs = numpy.zeros(3)
Base.fill_b(xs, 10)
assert all(ys == 10)
```

or

```python
xs = numpy.array([1, 2, 3])
ys = xs[:]
if julia.eval("VERSION >= v"0.7-""):
    Base.rmul_b(xs, 2)
else:
    Base.scale_b(xs, 2)
assert all(2 * ys == xs)
```

Which one do you like?

Reply:
Actually, the test I proposed does not work. Strange...

```python
    def test_julia_module_bang(self):
        from julia import Base
        xs = numpy.zeros(3)
        ys = Base.fill_b(xs, 10)
        assert all(ys == 10)
>       assert all(xs == 10)
E       AssertionError: assert False
E        +  where False = all(array([0., 0., 0.]) == 10)
```

Reply:
This happens in PyCall as well.  Is it a known issue or this is intended?

```julia
julia> py"""
       import numpy
       xs = numpy.zeros(3)
       """

julia> fill!(py"xs", 10)
3-element Array{Float64,1}:
 10.0
 10.0
 10.0

julia> py"xs"
3-element Array{Float64,1}:
 0.0
 0.0
 0.0

(v1.0) pkg> st
    Status `~/.julia/environments/v1.0/Project.toml`
  ...
  [438e738f] PyCall v1.18.0
  ...
```

Reply:
Ah, that was because `py"xs"` was converted to `PyAny`.  If I do `fill!(PyArray(PyCall.maindict()["xs"]), 10)` in Julia, mutation works (= it's reflected on Python side).

I'm not sure what's the right illustration of this problem in Julia.  Something like this?:

```julia
julia> py"""
       def apply(f):
           xs = numpy.zeros(3)
           f(xs)
           return xs
       """

julia> py"apply"(xs -> fill!(xs, 10))
3-element Array{Float64,1}:
 0.0
 0.0
 0.0
```

Reply:
See https://github.com/JuliaPy/PyCall.jl/pull/531#issuecomment-414314622 … you need to specify the type conversion you want for an argument if you want in-place operations to be possible in general.

Reply:
Got it. Yes, I think explicitly denoting which argument is mutating is a good idea.

So, for this test, we just need to check the returned value, right?:

```python
    def test_julia_module_bang(self):
        from julia import Base
        xs = [1, 2, 3]
        ys = Base.fill_b(xs, 10)
        assert all(y == 10 for y in ys)
```

Although it may be better if we just drop `_b` since users have to implement a callback on Julia side to avoid allocations.

Reply:
OK. So I think I finally found a meaningful test:

```python
    def test_julia_module_bang(self):
        from julia.Base import Channel, put_b, take_b
        chan = Channel(1)
        sent = 123
        put_b(chan, sent)
        received = take_b(chan)
        assert sent == received
```

What do you think?

In the future, I think we need an API to denote some Python values have to be passed without copying.  Something like `fill_b(asref(arr), 10)`.

Reply:
I think the right thing to do is to make it easy to use the `pyfunction` method from Julia, for cases where you need control over precisely how arguments are passed.